apps/src/blockly/addons/cdoBlockShadow.js

```diff
@@ -0,0 +1,141 @@
+// Helper functions for block shadowing logic. We use block shadowing for
+// our "mini toolbox" blocks, which are blocks that contain their own toolbox.
+// The blocks in the toolbox (or the blocks pulled from the toolbox)
+// need to shadow the image selected in the parent
+// block.
+
+// This object maps the block types that shadow parents to the parent block type they can shadow.
+// It also includes the index of the image to shadow in the parent's inputList.
+export const blockShadowingPairs = {
+  gamelab_clickedSpritePointer: {
+    parent: 'gamelab_spriteClicked',
+    imageIndex: 0,
+  },
+  gamelab_newSpritePointer: {
+    parent: 'gamelab_whenSpriteCreated',
+    imageIndex: 0,
+  },
+  gamelab_subjectSpritePointer: {
+    parent: 'gamelab_checkTouching',
+    imageIndex: 0,
+  },
+  gamelab_objectSpritePointer: {
+    parent: 'gamelab_checkTouching',
+    imageIndex: 1,
+  },
+};
+
+// Given a shadow block and an optional parent block id, return the url of the image
+// that the shadow block should display, or an empty string if the block should display
+// no image. If given a parent id, use that block as the parent, otherwise try to find the parent.
+// The parent block will either be the root block of the shadow block (if the shadow block is not inside
+// a mini toolbox), or the source block of the block's workspace (if the shadow block is inside a mini toolbox).
+export function getShadowedBlockImageUrl(block, parentId) {
+  const blockShadowData = blockShadowingPairs[block.type];
+  if (!blockShadowData || !block.inputList || block.inputList.length === 0) {
+    return '';
+  }
+  let parent = undefined;
+  if (parentId !== undefined) {
+    parent = Blockly.getMainWorkspace().getBlockById(parentId);
+  }
+  if (!parent) {
+    const rootBlock = block.getRootBlock();
+    if (rootBlock.type === blockShadowData.parent) {
+      parent = rootBlock;
+    } else if (block.getRootBlock().type === block.type) {
+      const workspace = Blockly.Workspace.getById(block.workspace.id);
+      // If this block has itself as a parent and is in a flyout workspace,
+      // it may be a in a mini toolbox. We check the parentBlockId on the block to
+      // confirm it is still inside the parent flyout workspace.
+      if (workspace && workspace.isFlyout && block.parentBlockId) {
+        const potentialParent = Blockly.getMainWorkspace().getBlockById(
+          block.parentBlockId
+        );
+        // Confirm potential parent is the correct type and it
+        // does contain this block's current workspace.
+        if (potentialParent.type === blockShadowData.parent) {
```

Comment:
The checks inside this if might be unnecessary. I don't know if there's a case where a block could be inside a flyout workspace that was not the original parent workspace. Happy to remove this if someone is more confident that we don't need it!

Reply:
makes sense, I'll remove this check